### PR TITLE
Fix credentials not refreshed in ds tree and misc profile related fixes

### PR DESCRIPTION
**Pull request by t1m0thyj**
## Proposed changes

This PR resolves several issues related to profile management in V3:
* #3112
* #2259
* #3134
* #3120

**How to Test**
Test that profile changes are propagated:
  * Update credentials (user/password)
  * Login/logout with base profile
  * Login/logout with regular profile
  * All the above when autoStore=false

## Release Notes

Milestone: 3.0.0

Changelog: TBD

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (non-breaking change which adds or improves functionality)
- [ ] Breaking change (a change that would cause existing functionality to not work as expected)
- [ ] Documentation (Markdown, README updates)
- [ ] Other (please specify above in "Proposed changes" section)

## Checklist

**General**

- [ ] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [ ] All PR dependencies have been merged and published (if applicable)
- [ ] A GIF or screenshot is included in the PR for visual changes
- [ ] The pre-publish command has been executed:
  - **v2 and below:** `yarn workspace vscode-extension-for-zowe vscode:prepublish`
  - **v3:** `pnpm --filter vscode-extension-for-zowe vscode:prepublish`

**Code coverage**

- [ ] There is coverage for the code that I have added
- [ ] I have added new test cases and they are passing
- [ ] I have manually tested the changes

**Deployment**

- [ ] I have added developer documentation (if applicable)
- [ ] Documentation should be added to Zowe Docs
  - If you're an outside contributor, please post in the [#zowe-doc Slack channel](https://openmainframeproject.slack.com/archives/CC961JYMQ) to coordinate documentation.
  - Otherwise, please check with the rest of the squad about any needed documentation before merging.
- [ ] These changes may need ported to the appropriate branches (list here):

### CODE REVIEW

**Review by JillieBeanSim — CHANGES_REQUESTED** [review]
Thanks for working on these bugs @t1m0thyj 
I did experience some strange behavior after updating credentials from failing one to working one and successfully did search and expand members but got the below error in text editor when trying to open file
Screenshot 2024-09-24 at 9 14 02 AM

**Comment by t1m0thyj** [comment]
@JillieBeanSim Thanks for catching this, I think I've fixed the issue but need to update some tests 😋

**Review by zFernand0 — COMMENTED** [review]
LGTM! 😋 

Left a couple of comments but nothing that requires changes on this PR 😋  

Nice to see the error icon for global profiles now, thank you!
image

A few quirks:
- With `autoStore: false` and only the login regular profile
  - ✅ triggering the equivalent of updateCreds works fine with nested profiles
  - ❓ triggering the equivalent of updateCreds with flat profiles throws an error, but works fine
    - Entering credentials for the first time in a flat config file displays an error message but the operation works and the credentials are propagated across ZE
      image
  - ✅ login with base profile works fine
  - ✅ logout works fine for regular and base
  - ❓ login with regular/nested profiles
    - I noticed that the same profile in other trees does prompt for credentials to login, however, it works fine if I escape that first prompt and go to the actual search. That means that the credentials are propagated, but perhaps the checkCurrentProfile (or similar method) is not aware of the new credentials.
- With `autoStore: true`:
  - Mostly the same as `autoStore: false`, but I ran into the pop-up frequently for the updareCreds approach
      image

File: packages/zowe-explorer-api/src/profiles/ProfilesCache.ts [hunk 1]
```diff
@@ -79,9 +80,8 @@ export class ProfilesCache {
 
     /**
      * Updates profile in allProfiles array and if default updates defaultProfileByType
-     *
+     * @deprecated Use `updateCachedProfile` instead
```
[the remark this answers is not in the document]
Comment (t1m0thyj) [reply packages/zowe-explorer-api/src/profiles/ProfilesCache.ts:83]: I don't think we can here because the old method is synchronous and the new one returns a Promise.

**Comment by t1m0thyj** [comment]
@zFernand0 Thanks for the detailed testing and feedback, I believe issues should be fixed now 😋

**Review by zFernand0 — APPROVED** [review]
issues with `autoStore: true` have been resolved.

However, I feel autoStore: `false` isn't ideal yet 😋 
- Credentials are not propagated between trees (which I believe is expected)
- Login is propagated (even though you have to escape the login prompt), but logout is not.
  - Logout happens to remove the credential propagation, i.e. I can't navigate inside folders, but it lets you navigate (perhaps from cached contents) to the same directory from before (even though the token should be expired)
  
I know `autoStore: false` is not the most common configuration, and the things I mentioned here are small enough that people should not even encounter unless they are purposely trying to get into a weird state. 😋 

TL;DR: LGTM! 😋

**Review by JillieBeanSim — APPROVED** [review]

**Review by traeok — COMMENTED** [review]
I'm still running into a 401 error when opening PDS members - here's the scenario I tested:

1. Enter a search pattern for one of your profiles with pre-existing, valid credentials in place
2. With the session expanded and data sets listed, right-click on profile node and select Manage Profile -> Update Credentials
3. Enter in invalid credentials for the profile
4. When the "invalid credentials" prompt occurs for the profile, cancel out of the prompt
  a. Notice that the children elements for the profile node are removed
5. Right click on profile node -> Manage Profile -> Update Credentials
6. Enter in valid credentials this time
7. Click on profile node to toggle its collapsible state to expanded
8. Open one of the PDS in the list
9. Click on one of its members

I realized this was merged in the middle of my review, but figured I should post it anyway as it will likely need addressed in a future PR. The scenario is a bit niche, but could be reproduced if a user doesn't update their invalid creds right away.

image

**Comment by JillieBeanSim** [comment]
@t1m0thyj can a bug report be created for the edge case please? TIA
